examples/arithmetics/src/language-server/arithmetics-validator.ts

```diff
@@ -93,4 +137,59 @@ export class ArithmeticsValidator {
             accept('error', `Function ${functionCall.func.ref?.name} expects ${functionCall.args.length} parameters, but ${(functionCall.func.ref as Definition).args.length} were given.`, { node: functionCall, property: 'args' });
         }
     }
-}
\ No newline at end of file
+}
+
+type FunctionCallTree = Map<ResolvedFunctionCall, NestedFunctionCall>
+type FunctionCallCycle = NestedFunctionCall[]
+namespace FunctionCallCycle {
+
+    export function select(to: NestedFunctionCall, tree: FunctionCallTree): FunctionCallCycle | undefined {
+        const referencedFunc = to.call.func.ref;
+        let parent = tree.get(to.call);
+        while (parent) {
+            if (parent.host === referencedFunc) return [parent, to];
+            parent = tree.get(parent.call);
+        }
+        return undefined;
+    }
+
+    export function print(cycle: FunctionCallCycle, tree: FunctionCallTree): string {
+        return stream(iterateBack(cycle, tree))
+            .map(NestedFunctionCall.toString)
+            .reduce((child, parent) => parent + '->' + child) ?? '';
+    }
+
+    export function* iterateBack(cycle: FunctionCallCycle, tree: FunctionCallTree): Generator<NestedFunctionCall> {
+        const start = cycle[0];
+        const end = cycle[cycle.length - 1];
+        yield end;
+        if (start === end) return;
+        let parent = tree.get(end.call);
+        while (parent) {
+            yield parent;
+            if (parent.call === start.call) break;
+            parent = tree.get(parent.call);
+        }
+    }
+}
+
+type NestedFunctionCall = {
+    call: ResolvedFunctionCall,
+    host: Definition
+}
+namespace NestedFunctionCall {
+
+    export function* selectCalls(host: Definition, expression: Expression = host.expr): Generator<NestedFunctionCall> {
+        if (isFunctionCall(expression)) {
+            if (isResolvedFunctionCall(expression)) yield { call: expression, host };
+        } else if (isBinaryExpression(expression)) {
+            for (const expr of [expression.left, expression.right]) {
+                if (expr) yield* selectCalls(host, expr);
+            }
+        }
+    }
+
+    export function toString({ host, call }: NestedFunctionCall): string {
+        return `${host.name}::${call.func.ref.name}()`;
```

Comment:
This `::` syntax is not used in the DSL, so its meaning could be harder to understand; I would simplify it by removing the "host".

Reply:
Agree, the meaning can be inferred from the previous call in `fun1()->fun2()->fun3()` cycle.